src/VisualStudio/Core/Def/Implementation/Snippets/AbstractSnippetExpansionClient.cs

```diff
@@ -393,7 +474,361 @@ public virtual bool TryInsertExpansion(int startPositionInSubjectBuffer, int end
+            return semanticInfo.ReferencedSymbols;
+        }
+
+        /// <summary>
+        /// Update the current argument value completion session to use a specific method.
+        /// </summary>
+        /// <param name="method">The currently-selected method in Signature Help.</param>
+        /// <param name="cancellationToken">A cancellation token the operation may observe.</param>
+        public void MoveToSpecificMethod(IMethodSymbol method, CancellationToken cancellationToken)
+        {
+            AssertIsForeground();
+
+            if (ExpansionSession is null)
+            {
+                return;
+            }
+
+            if (SymbolEqualityComparer.Default.Equals(_state._method, method))
+            {
+                return;
+            }
+
+            var symbolName = _state._method?.Name ?? _state._methods.FirstOrDefault()?.Name;
+            if (symbolName != method.Name)
+            {
+                // Signature Help is showing a signature that wasn't part of the set this argument value completion
+                // session was created from. It's unclear how this state should be handled, so we stop processing
+                // Signature Help updates for the current session.
+                // TODO: https://github.com/dotnet/roslyn/issues/50636
+                _state.ClearSymbolInformation();
+                return;
+            }
+
+            if (_state._methods.IsDefaultOrEmpty)
+            {
+                // Signature Help is showing a set of overloads that don't match the overloads from the point where the
+                // argument completion session first started. It's unclear how this state should be handled, so we stop
+                // processing Signature Help updates for the current session.
+                // TODO: https://github.com/dotnet/roslyn/issues/50636
+                _state.ClearSymbolInformation();
+                return;
+            }
+
+            var textViewModel = TextView.TextViewModel;
+            if (textViewModel == null)
+            {
+                Debug.Assert(TextView.IsClosed);
+                return;
+            }
+
+            var buffer = EditorAdaptersFactoryService.GetBufferAdapter(textViewModel.DataBuffer);
+            if (buffer is not IVsExpansion expansion)
+            {
+                return;
+            }
+
+            // Track current argument values so input created/updated by a user is not lost when cycling through
+            // Signature Help overloads:
+            //
+            // 1. For each parameter of the method currently presented as a snippet, the value of the argument as
+            //    it appears in code.
+            // 2. Place the argument values in a map from parameter name to current value.
+            // 3. (Later) the values in the map can be read to avoid providing new values for equivalent parameters.
+            if (_state._method is not null)
```

Comment:
Any concern we're holding onto methods during the lifetime of the snippet and rooting old compilations/Solutions?

Reply:
➡️ I'm not currently concerned. All of this will be rewritten as part of supporting nested calls anyway.

Reply:
When is that coming?

Reply:
No specific ETA, but having using this feature it will be an early and common request.

Overall this is a short-lived action and we only root solutions while the snippet is active. Another reason to not be concerns is we had a bug where we _permanently_ rooted a solution for every open text view as part of Quick Info for completion lists, and we went a few releases before anyone even noticed.